zerver/tests/test_previewable.py

```diff
@@ -0,0 +1,212 @@
+                "html_url": "https://github.com/zulip/zulip/pull/22368",
+                "diff_url": "https://github.com/zulip/zulip/pull/22368.diff",
+                "patch_url": "https://github.com/zulip/zulip/pull/22368.patch",
+                "merged_at": None,
+            },
+        }
+        responses.add(
+            responses.GET,
+            "https://api.github.com/repos/zulip/zulip/issues/22368",
+            json=response_data,
+        )
+        result = self.client_post(
+            "/json/previewable", {"url": "https://github.com/zulip/zulip/pull/22368"}
+        )
+        self.assert_json_success(result)
+        result = result.json()
+        self.assertEqual(result["platform"], "github")
+        self.assertEqual(result["type"], "pulls")
+        self.assertEqual(result["owner"], "zulip")
+        self.assertEqual(result["repo"], "zulip")
+        self.assertEqual(result["issue_number"], "22368")
+        self.assertEqual(result["state"], "open")
+        self.assertEqual(result["user"], {"login": "brijsiyag"})
+        self.assertEqual(
+            result["title"], "Feat: Show titles of GitHub issues on hover fixes: #19710."
+        )
+        self.assertIsNotNone(result.get("pull_request", None))
+        self.assertEqual(
+            result["pull_request"],
+            {
+                "url": "https://api.github.com/repos/zulip/zulip/pulls/22368",
+                "html_url": "https://github.com/zulip/zulip/pull/22368",
+                "diff_url": "https://github.com/zulip/zulip/pull/22368.diff",
+                "patch_url": "https://github.com/zulip/zulip/pull/22368.patch",
+                "merged_at": None,
+            },
+        )
+
+    def test_not_previewable_url(self) -> None:
+        result = self.client_post(
+            "/json/previewable", {"url": "https://github.com/zulip/xyz/zulip/issues/3"}
+        )
+        self.assert_json_error(result, msg="URL is not previewable.")
+
+    def test_not_url(self) -> None:
+        result = self.client_post(
+            "/json/previewable", {"url": "https://github.com[/zulip/xyz/zulip/issues/3"}
+        )
+        self.assert_json_error(result, msg="URL is not valid.")
+
+    @responses.activate()
+    def test_invalid_previewable_url(self) -> None:
+        response_data = (
+            {
+                "message": "Not Found",
+                "documentation_url": "https://docs.github.com/rest/reference/issues#get-an-issue",
+            },
+        )
+        api_url = "https://api.github.com/repos/zulip/zulip/issues/2"
+        responses.add(
+            responses.GET,
+            api_url,
+            json=response_data,
+            status=404,
+        )
+        response = self.client_post(
+            "/json/previewable", {"url": "https://github.com/zulip/zulip/issues/2"}
+        ).json()
+        self.assertEqual(response["msg"], "Invalid or expired URL.")
```

Comment:
I think we should use `self.assert_json_error` here, so we also check the status code.